examples/hackweek_demo/plot_pnl.py

```diff
@@ -112,93 +113,134 @@ def get_wallet_from_onchain_trade_info(
+    """Arrange agent wallet info for plotting, while calculating open share price."""
+    if session is None:
+        session = postgres.initialize_session()
+    all_agent_wallet_info = postgres.get_wallet_info_history(session)
+    agent_info = {}
+    for agent, wallet in all_agent_wallet_info.items():
+        share_price = wallet["sharePrice"].rename({"sharePrice": "share_price"})
+
+        # Create positions
+        positions = wallet.loc[:, wallet.columns[:-2]].copy()
+        positions = positions.where(positions != 0, pd.NA)
+
+        # Create deltas
+        deltas = positions.diff()
+        deltas.iloc[0] = positions.iloc[0]
+
+        # Create NaNs of the same size as deltas
+        share_price_on_increases = pd.DataFrame(data=pd.NA, index=deltas.index, columns=deltas.columns)
+
+        # Replace positive deltas with share_price
+        share_price_on_increases = share_price_on_increases.where(deltas <= 0, share_price, axis=0)
+
+        # Fill forward to replace NaNs, to have updated share prices only on position increases
+        share_price_on_increases.fillna(method="ffill", inplace=True, axis=0)
+
+        # Calculate weighted average share price across all deltas
+        # weighted_average_share_price = (share_price_on_increases * deltas).cumsum(axis=0) / positions
+        weighted_average_share_price = pd.DataFrame(data=pd.NA, index=deltas.index, columns=deltas.columns)
+        weighted_average_share_price.iloc[0] = share_price_on_increases.iloc[0]
+        for row in deltas.index[1:]:
+            # index positive deltas
+            cols = deltas.loc[row, :] > 0
+
+            new_avg = []
+            if len(cols) > 0:
+                # calculate update
+                # new_avg = (old_amount * old_avg + delta_amount * delta_avg) / (old_amount + delta_amount)
+                new_avg = (
+                    share_price_on_increases.loc[row, cols] * deltas.loc[row, cols]
+                    + weighted_average_share_price.loc[row - 1, cols] * positions.loc[row - 1, cols]
+                ) / (deltas.loc[row, cols] + positions.loc[row - 1, cols])
+
+            # keep previous result where delta isn't positive, otherwise replace with new_avg
+            weighted_average_share_price.loc[row, :] = weighted_average_share_price.loc[row - 1, :].where(
+                ~cols, new_avg, axis=0
             )
 
+        agent_info[agent] = Info(
+            wallet=wallet,
+            positions=positions,
+            deltas=deltas,
+            open_share_price=share_price_on_increases,
+        )
+    return agent_info
+
+
+def calculate_pnl(pool_config, pool_info, checkpoint_info):
+    """Calculate pnl for all agents."""
+    #  pylint: disable=too-many-locals
+    session = postgres.initialize_session()
+    all_agent_info = get_agent_info(session)
+    position_duration = pool_config.positionDuration.iloc[0]
+
+    for _, agent_info in all_agent_info.items():
```

Comment:
`for agent_info in all_agent_info.values()`?

Reply:
kek forgot that exists